src/ucp/rndv/rndv_rkey_ptr.c

```diff
@@ -183,3 +190,266 @@ ucp_proto_t ucp_rndv_rkey_ptr_proto = {
+        .super.latency       = 0,
+        .super.cfg_thresh    = ucp_proto_rndv_cfg_thresh(context, rndv_modes),
+        .super.cfg_priority  = 0,
+        .super.min_length    = 0,
+        .super.max_length    = max_length,
+        .super.min_iov       = 1,
+        .super.min_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.put.min_zcopy),
+        .super.max_frag_offs = ucs_offsetof(uct_iface_attr_t,
+                                            cap.put.max_zcopy),
+        .super.max_iov_offs  = ucs_offsetof(uct_iface_attr_t, cap.put.max_iov),
+        .super.hdr_size      = 0,
+        .super.send_op       = UCT_EP_OP_LAST,
+        .super.memtype_op    = UCT_EP_OP_GET_ZCOPY,
+        .super.flags         = UCP_PROTO_COMMON_INIT_FLAG_RKEY_PTR,
+        .max_lanes           = context->config.ext.max_rndv_lanes,
+        .initial_reg_md_map  = initial_reg_md_map,
+        .first.tl_cap_flags  = 0,
+        .first.lane_type     = UCP_LANE_TYPE_RKEY_PTR,
+        .middle.tl_cap_flags = 0,
+        .middle.lane_type    = UCP_LANE_TYPE_RKEY_PTR
+    };
+    size_t bulk_priv_size;
+    ucs_status_t status;
+
+    status = ucp_proto_rndv_bulk_init(&params, &rpriv->bulk,
+                                      UCP_PROTO_RNDV_RKEY_PTR_DESC,
+                                      UCP_PROTO_RNDV_ATP_NAME, &bulk_priv_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    rpriv->atp_map       = rpriv->bulk.mpriv.lane_map;
+    rpriv->atp_num_lanes = ucs_popcount(rpriv->atp_map);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_proto_rndv_rkey_ptr_mtype_init(const ucp_proto_init_params_t *init_params)
+{
+    ucp_context_t *context = init_params->worker->context;
+    ucs_status_t status;
+    ucp_md_map_t mdesc_md_map;
+    size_t frag_size;
+
+    ucp_md_index_t md_index;
+    const uct_md_attr_t *md_attr;
+    int found_md_map = 0;
+
+    if (!context->config.ext.rndv_shm_ppln_enable) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if ((init_params->select_param->dt_class != UCP_DATATYPE_CONTIG) ||
+        !ucp_proto_rndv_op_check(init_params, UCP_OP_ID_RNDV_SEND, 1) ||
+        init_params->rkey_config_key == NULL) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    status = ucp_proto_rndv_mtype_init(init_params, &mdesc_md_map, &frag_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    ucs_for_each_bit(md_index, init_params->rkey_config_key->md_map) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if ((md_attr->cap.flags & UCT_MD_FLAG_RKEY_PTR) &&
+            /* Do not use xpmem, because cuda_copy registration will fail and
+             * performance will not be optimal. */
+            !(md_attr->cap.flags & UCT_MD_FLAG_REG) &&
+            (md_attr->cap.access_mem_types &
+             UCS_BIT(init_params->rkey_config_key->mem_type))) {
+            found_md_map = UCS_BIT(md_index);
```

Comment:
we will need to take into account remote md indicies rather than local. Will fix it in proto v1 and then we can do the same here